agent/consul/discoverychain/gateway_httproute.go

```diff
@@ -0,0 +1,276 @@
+// https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.HTTPRouteRule
+//
+// Ordering prefers matches based on the largest number of:
+//
+//  1. characters in a matching non-wildcard hostname
+//  2. characters in a matching hostname
+//  3. characters in a matching path
+//  4. header matches
+//  5. query param matches
+//
+// The hostname-specific comparison (1+2) occur in Envoy outside of our control:
+// https://github.com/envoyproxy/envoy/blob/5c4d4bd957f9402eca80bef82e7cc3ae714e04b4/source/common/router/config_impl.cc#L1645-L1682
+func compareHTTPRules(ruleA, ruleB structs.HTTPMatch) bool {
+	if len(ruleA.Path.Value) != len(ruleB.Path.Value) {
+		return len(ruleA.Path.Value) > len(ruleB.Path.Value)
+	}
+	if len(ruleA.Headers) != len(ruleB.Headers) {
+		return len(ruleA.Headers) > len(ruleB.Headers)
+	}
+	return len(ruleA.Query) > len(ruleB.Query)
+}
+
+func httpServiceDefault(entry structs.ConfigEntry, meta map[string]string) *structs.ServiceConfigEntry {
+	return &structs.ServiceConfigEntry{
+		Kind:           structs.ServiceDefaults,
+		Name:           entry.GetName(),
+		Protocol:       "http",
+		Meta:           meta,
+		EnterpriseMeta: *entry.GetEnterpriseMeta(),
+	}
+}
+
+func synthesizeHTTPRouteDiscoveryChain(route structs.HTTPRouteConfigEntry) (structs.IngressService, *structs.ServiceRouterConfigEntry, []*structs.ServiceSplitterConfigEntry, []*structs.ServiceConfigEntry) {
+	meta := route.GetMeta()
+	splitters := []*structs.ServiceSplitterConfigEntry{}
+	defaults := []*structs.ServiceConfigEntry{}
+
+	router, splits := httpRouteToDiscoveryChain(route)
+	serviceDefault := httpServiceDefault(router, meta)
+	defaults = append(defaults, serviceDefault)
+	for _, split := range splits {
+		splitters = append(splitters, split)
+		if split.Name != serviceDefault.Name {
+			defaults = append(defaults, httpServiceDefault(split, meta))
+		}
+	}
+
+	ingress := structs.IngressService{
+		Name:           router.Name,
+		Hosts:          route.Hostnames,
+		Meta:           route.Meta,
+		EnterpriseMeta: route.EnterpriseMeta,
+	}
+
+	return ingress, router, splitters, defaults
+}
+
+func httpRouteToDiscoveryChain(route structs.HTTPRouteConfigEntry) (*structs.ServiceRouterConfigEntry, []*structs.ServiceSplitterConfigEntry) {
+	router := &structs.ServiceRouterConfigEntry{
+		Kind:           structs.ServiceRouter,
+		Name:           route.GetName(),
+		Meta:           route.GetMeta(),
+		EnterpriseMeta: route.EnterpriseMeta,
+	}
+	var splitters []*structs.ServiceSplitterConfigEntry
+
+	for idx, rule := range route.Rules {
+		modifier := httpRouteFiltersToServiceRouteHeaderModifier(rule.Filters.Headers)
+		prefixRewrite := httpRouteFiltersToDestinationPrefixRewrite(rule.Filters.URLRewrites)
+
+		var destination structs.ServiceRouteDestination
+		if len(rule.Services) == 1 {
+			// TODO open question: is there a use case where someone might want to set the rewrite to ""?
```

Comment:
Doesn't have to be resolved in this PR, but I believe the answer to this is "yes".

In previous places I've worked, you would do something like target a specific microservice in the mesh using `GET https://domain.com/service/myservice/v1/resources`.

`myservice` would serve the route `/v1/resources` but had no need to know how your request got to that service, so you would match on `/service/myservice` and send it to the `myservice` backend while rewriting the prefix to `""`.

This is just one use case that's probably pretty edge-casey, but it demonstrates why you might do such a thing. IIRC we also need to support this to fulfill the K8s Gateway API.

Reply:
Sounds good, I'll loop around and handle this as we do cleanup.